Source/Diagnostics/Diagnostics.cpp

```diff
@@ -222,7 +222,7 @@ Diagnostics::InitData ()
         // Note that the filter is set for every ith snapshot, and the number of snapshots
         // for full diagnostics is 1, while for BTD it is user-defined.
         for (int i_buffer = 0; i_buffer < m_num_buffers; ++i_buffer ) {
-            for (int i = 0; i < m_output_species.size(); ++i) {
+            for (int i = 0; i < m_output_species[i_buffer].size(); ++i) {
```

Comment:
This might introduce the same bug as #2738 and needs some extra check?

Reply:
Ah I see. Yes! this is true!
Thanks for catching it!
WIll fix this

Reply:
The latest commit should fix it?
Do you have an input file that triggered this Backtrace in #2738

Reply:
Ah never mind. I found the input file, as was mentioned in #2738  .
I ran this case locally and it works now.

Reply:
Thank you!

Can you please replace the unchecked `operator[]` with `.at()`? This would make this way safer, because of a runtime check.

I cannot fully understand why this part would work. `m_num_buffers` is not zero in this case, but the size of `m_output_species` is. The access of `m_output_species[0]` is then undefined behavior (a memory violation).

Reply:
Thank you @ax3l . I just realized why you mentioned this. 
Now when write_species is zero, species output is cleared for the ith buffer, but the whole 2D species_output vector is not cleared.

I have run this with DEBUG and also used` .at()`